### PR TITLE
Filter partial calls

### PR DESCRIPTION
**Pull request by timothymillar**
Issue #223 

Some notes:

- Adding a new top-level sub-module might be OTT but this doesn't really fit in `stats` or `model`.
- Currently this will always throw a merge warning which is a safe default but may be inconvenient. Is an argument to optionally silence merge warnings a good idea?
- This implementation will require some minor changes if/when #243 is resolved.
- Some discussion in issue #282 on handling partial genotypes in other functions.

### CODE REVIEW

**Comment by timothymillar** [comment]
Handling of warnings related to #288

File: sgkit/curate.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    Dataset containing updated variables `call_genotype` and
+    `call_genotype_mask` in which partial genotype calls are replaced
+    with compleately missing genotype calls.
+
+    Examples
+    --------
+
+    >>> import warnings
+    >>> import sgkit as sg
+    >>> from sgkit.testing import simulate_genotype_call_dataset
```
Comment (tomwhite) [inline sgkit/curate.py:43]: `simulate_genotype_call_dataset` is in the top-level namespace now, so you can just reference as `sg. simulate_genotype_call_dataset`

File: sgkit/curate.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+from typing import Hashable
```
Comment (tomwhite) [inline sgkit/curate.py:1]: Note sure about the name of this file. Another possibility might be `preprocessing.py`?

**Review by tomwhite — COMMENTED** [review]
> Currently this will always throw a merge warning which is a safe default but may be inconvenient. Is an argument to optionally silence merge warnings a good idea?

It does seem odd that this will always produce a merge warning, even though it's the expected behaviour. I think that will be confusing for users.

It's worth pointing out that the dataset that is passed in will not be updated, so the unfiltered calls are still available if needed. So this makes me think that the default behaviour here should probably *not* be to throw the warning, but I'm not sure how to make this consistent with other methods.

Thoughts @eric-czech, @jeromekelleher, @ravwojdyla?

**Comment by eric-czech** [comment]
Hey @timothymillar, we were talking about this on a call today and we discussed a few options:

1. Making new variables for all functions that perform some kind of mutation (e.g. `call_genotype_complete[_mask]` here maybe)
2. Make an exception for mutation functions and have them only return new variables instead of merging
3. Disable the warning 
4. Rename the original variable as well as the new variable 

The consensus we landed on was option 1.  The advantages of this are:

- Multiple mutation functions relying on the same original variable can be run in sequence without the order mattering
- This can be more consistent with mutation functions that alter the interpretation of the variable in such a way that we would likely want to create a new sgkit.variable for the result anyhow, e.g. mean imputation.
- It doesn't alter the input variables as all other functions do

The biggest disadvantage is that downstream functions expecting to work on `call_genotype` need to be pointed at something else, which won't be obvious to the user.  For example:

```python
ds.pipe(sg.filter_partial_calls)\
    .pipe(sg.count_call_alleles, call_genotype=sg.variables.call_genotype_complete)
```

We aren't thrilled about that but it seems like the lesser of several evils.  How does that sound to you?

File: sgkit/curate.py (hunk 2 above)
Comment (timothymillar) [reply sgkit/curate.py:1]: Agreed, `preprocessing` is better

**Comment by timothymillar** [comment]
@eric-czech @tomwhite thanks for the feedback. 
I'm happy with option 1, it is the most consistent approach. Though it would require some clear documentation to ensure that users don't create `call_genotype_complete` and then use `call_genotype` without realizing.

**Comment by timothymillar** [comment]
Sorry for the delay on this (busy week/new dev machine).

For the example code, it is also possible to drop and rename via xarray methods:
```python
ds2 = (
    sg.filter_partial_calls(ds)
    .drop(['call_genotype', 'call_genotype_mask'])
    .rename(call_genotype_complete='call_genotype', call_genotype_complete_mask='call_genotype_mask')
)
```
However my thought here is that simple overwriting syntax is probably more familiar to most users.

File: sgkit/variables.py [hunk 3]
```diff
@@ -176,6 +176,20 @@ def _check_field(
 A flag for each call indicating if it is phased or not. If omitted
 all calls are unphased.
 """
+call_genotype_complete, call_genotype_complete_spec = SgkitVariables.register_variable(
+    ArrayLikeSpec("call_genotype_complete", kind="i", ndim=3)
+)
+"""
+Call genotypes in which partial genotype calls are replaced with
+compleately missing genotype calls.
```
Comment (eric-czech) [inline sgkit/variables.py:184]:
```suggestion
completely missing genotype calls.
```

File: sgkit/stats/preprocessing.py [hunk 4]
```diff
@@ -102,3 +107,77 @@ def inverse_transform(self, X: ArrayLike, copy: Optional[bool] = None) -> ArrayL
...
+        If True (the default), merge the input dataset and the computed
+        output variables into a single dataset, otherwise return only
+        the computed output variables.
+        See :ref:`dataset_merge` for more details.
+
+    Returns
+    -------
+    Dataset containing `call_genotype_complete` and
+    `call_genotype_complete_mask` in which partial genotype calls are
+    replaced with compleately missing genotype calls.
```
Comment (eric-czech) [inline sgkit/stats/preprocessing.py:138]:
```suggestion
    replaced with completely missing genotype calls.
```

**Review by eric-czech — APPROVED** [review]

**Comment by eric-czech** [comment]
Would you like to take another look at this one @tomwhite?  I was going to set it to merge if not.

**Review by tomwhite — APPROVED** [review]

**Comment by eric-czech** [comment]
Thanks @tomwhite, I'll set it to merge once those two tiny misspelling suggestions are committed @timothymillar.

**Pull request merged**